blog/tech-talk-window-resize-behavior.md

```diff
@@ -0,0 +1,290 @@
+
+By chance, I stumbled across a few helpful debugging flags in the Chromium source code:
+
+- `--ui-show-paint-rects`
+- `--ui-show-property-changed-rects`
+- `--ui-show-surface-damage-rects`
+- `--ui-show-composited-layer-borders`
+- `--tint-composited-content`
+- `--tint-composited-content-modulate`
+- (And more)
+
+They highlight areas of the browser window that were redrawn or updated by different parts of the Chromium graphics stack.
+
+That allowed me to see which part of the graphics stack was producing which output.
+
+In particular, the combination of `--tint-composited-content` and `--tint-composited-content-modulate` was really helpful. The former adds a tint to the output of the compositor. The latter changes the tint color on every frame.
+
+![Screenshot of Chromium with the --tint-composited-content flag](/assets/img/blog/tech-talk-window-resize-behavior/tinting-composited-content.png)
+
+In the screenshot, the cyan-tinted frame was the last frame that was being drawn.
+
+The jank to the right of that frame was not tinted cyan. It was tinted in different colors that were still there from previous frames. This indicated that the jank was not coming from the compositor. The compositor was sending the right output.
+
+The compositor is part of Chromium's graphics stack. The following is very simplified, but for the purpose of this blog post you can imagine it like this:
+
+1. The compositor `cc` produces a `CompositorFrame`, which contains draw instructions.
+2. `cc` sends that `CompositorFrame` to the display compositor `viz`.
+3. `viz` then draws the frame and shows it on screen.
+
+Tinting each `CompositorFrame` showed that the compositor produced the right output. So the issue had to be in the display compositor `viz`.
+
+## Locating the relevant `viz` code
+
+From there, I started searching for mentions of Direct3D in the `viz` source code.
+
+_Note: From here on, the post will get a bit more technical and reference source code symbols._
+
+I found that on the ANGLE Direct3D 11 backend, Chromium uses the Windows [DirectComposition](https://learn.microsoft.com/en-us/windows/win32/directcomp/directcomposition-portal) API for drawing the window contents.
+
+DirectComposition differs from all other output devices in Chromium. The DirectComposition `OutputSurface` has the capability `supports_viewporter` ([source link 1](https://source.chromium.org/chromium/chromium/src/+/main:ui/gl/dcomp_presenter.cc;l=144-146;drc=1cffe9643e1e70f1f369bbb28e015d9c5f968546), [source link 2](https://source.chromium.org/chromium/chromium/src/+/main:components/viz/service/display/output_surface.h;l=100-103;drc=c96a878b1cb45f60aac2285ffbdbd6b53dc92415)).
+
+An output surface is a bitmap that can be drawn to, often backed by a GPU texture.
+
+Without `supports_viewporter`, whenever the window size changes, Chromium will create a new output surface matching the new window size. Then it will draw on that surface and show it.
```

Comment:
```suggestion
Without `supports_viewporter`, Chromium will create a new output surface matching the new window size whenever the window size changes. Then it will draw on that surface and show it.
```

Shifting the clause here to make the sentence roll a bit smoother

Reply:
Tbh I'd prefer to keep it as it. I think it makes it easier to understand.

Here the relationship is "A leads to B". It's easier for the brain to understand in that order than if we mention B first and then bring up the condition for it, A. (See [here](https://web.archive.org/web/20250116041707/https://dilbertblog.typepad.com/the_dilbert_blog/2007/06/the_day_you_bec.html).)

I'd sacrifice smoothness for easier understanding here, but I see where you're coming from.